health/health.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package health
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/projectcalico/felix/set"
+)
+
+// Any kind of value that can be used as a map key and is unique across multiple packages.  For
+// example, "type myHealthSource string".
+type HealthSource interface{}
+
+type HealthIndicator struct {
+	// The source of this health indicator.
+	Source HealthSource
+
+	// How long the indicator is valid for.  In other words, if it continues operating normally,
+	// the source expects to refresh this indicator before this timeout.
+	Timeout time.Duration
+}
+
+// For a component that provides health indications, return the sources that it provides to indicate
+// readiness, and those that it provides to indicate liveness.
+type HealthProvider interface {
+	ReadySources() []HealthSource
+	LiveSources() []HealthSource
+}
+
+func MonitorHealth(
+	ready *bool,
```

Comment:
I think this would be cleaner as a method on a struct rather than a function with lots of parameters.  Then you have a place to document the mutex and what it protects, and you avoid needing to pass in pointers to bools, which is an odd pattern.